corgi/web/templates/product_manifest.json

```
@@ -1,4 +1,11 @@
 {# ProductModel manifest, obj is a ProductModel subclass here #}{% extends "base_manifest.json" %}{% load base_extras %}
+{% block document %}
+  "documentDescribes": [
+    "SPDXRef-{{external_name}}"
+  ],
+  "documentNamespace": "https://access.redhat.com/security/data/sbom/beta/spdx/{{external_name|escapejs}}",
```

The last part of the documentNamespace (the filename part) either SHOULD or MUST match either the actual filename, or the documentDescribes SPDXRef, or both, I believe. I think it's safer to just replace obj.name with external_name and not make any other changes.
```suggestion
  "documentNamespace": "https://access.redhat.com/security/data/sbom/beta/spdx/{{external_name|escapejs}}-{{obj.uuid}}",
```

If you really want to make this change as part of SBOM GA, please just double-check the spec quickly and maybe also run the online validator against a few sample manifests you generate:
https://tools.spdx.org/app/validate/

---

I did take a look at the spec but didnt' find anything indicating we need a uuid in the filename. Good idea, I will check the validator.

---

https://spdx.github.io/spdx-spec/v2.3/document-creation-information/#65-spdx-document-namespace-field says:

> The URI shall be unique for the SPDX document including the specific version of the SPDX document. If the SPDX document is updated, thereby creating a new version, a new URI for the updated document shall be used. There may only be one URI for an SPDX document and only one SPDX document for a given URI.

We comply with this by saying "we create new versions only when data is migrated from stage to prod". Only major updates create "new versions" of our documents with new UUIDs. This isn't really what the spec intends but is the most reasonable solution for us today.

If you do open that separate PR for only updating the SBOMs when the content has changed, we could start adding a random UUIDv4 to the filename and documentNamespace instead. SDEngine would then need to unpublish the old manifest before publishing the new one.

https://spdx.github.io/spdx-spec/v2.3/document-creation-information/#652-intent says:

> Although it is not required, the URI can be constructed in a way which provides information on how the SPDX document can be found. For example, the URI can be a URL referencing the SPDX document itself, if it is available on the internet.

> A best practice for creating the URI for SPDX documents available on the public internet is `https://[CreatorWebsite]/[pathToSpdx]/[DocumentName]-[UUID]`

Having the UUID in the documentNamespace, and making that match the filename, is only a "best practice". It's not even a SHOULD suggestion, nevermind a MUST requirement, so I think it's fine to remove the UUID from both places.

However, we need to be consistent with the component-based manifests. We can remove UUIDs from the filename and documentNamespace for both types of manifests.

Or we can generate a manifest for every ProductStream, if its external name has not already been processed (not part of the set). But we can still keep the UUID (for the first external name / stream we analyzed) as part of the documentNamespace and the filename, even if we don't consider UUIDs when figuring out "which streams / external names already generated a manifest?"

---

The published [ubi component manifest](https://access.redhat.com/security/data/sbom/beta/spdx/) don't include the UUID in the file name. [SDengine then builds the filename from the repository_url and version](https://gitlab.cee.redhat.com/prodsec-dev/sdengine/-/blob/c48ea029bc4f7f1b4341a47a39321764db1e7786/sdengine/tasks.py#L1580). Therefore we should do the same for the DocumentNamespace for component manifests, perhaps even expose an external_name attribute of the component on the API.

---

Yep, if we don't have this in the UBI / other component manifests, then removing it from product manifests as well sounds good to me. As long as we're consistent :)